### ISSUE TITLE
HybridBlocks don't work with constant NDArray initialized Parameters

### ISSUE THREAD

**Issue by Ajoo** [apache/incubator-mxnet#8438]
Hi. HybridBlocks with parameters initialized with a constant NDArray parameter don't seem to work with hybridize(). I always get a "TypeError: Object of type 'NDArray' is not JSON serializable" whan running hybridize() or tojson() on a symbol passed through.

I guess a minimal example to reproduce this error would be:

    import mxnet as mx

    class LinearMap(mx.gluon.HybridBlock):
        def __init__(self, A, **kwargs):
            super(LinearMap, self).__init__(**kwargs)
            with self.name_scope():
                self.A = self.params.get('A', shape=A.shape, init=mx.init.Constant(A))

        def hybrid_forward(self, F, x, A=None):
            return F.dot(A, x)
    
    l = LinearMap(mx.nd.ones((2,2)))
    l.initialize()

    l(mx.nd.array([1., 2.])) #this runs fine
    l.hybridize()
    l(mx.nd.array([1., 2.])) #this throws a TypeError

The same example runs fine when the initializer is replaced with any other, I'm guessing it just can't serialize the NDArray used to initialize with a constant.

As an aside, I didn't know that hybrid_forward would be called with parameters passed as keyword arguments. None of the tutorials mention it and the API docs aren't very explicit about it... I don't know if this is how hybrid_forward is supposed to be used.

**Comment by piiswrong** [comment]
Instead of using `init.Constant`, try `self.params.get('A', shape=A.shape, grad_req='null')` and set it to a fixed value after initialization `l.A.set_data(xx)`

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by lanking520** [comment]
Hi @Ajoo , are you still facing the same issue after tried the proposed solution?

**Comment by sernaton** [comment]
I have not been able to make the workaround suggested by @piiswrong work. In any case, it is a workaround. This is quite a bug! Constant initialization does not work with HybridBlocks....

**Comment by sernaton** [comment]
correction: the workaround did work after some tweaking, An example:
```
filter_mask = ... # an nd.array
self.filter_mask = self.params.get( 'filter_mask', shape=filter_mask.shape, grad_req='null' )
            self.filter_mask.initialize()
            self.filter_mask.set_data( filter_mask )
```
Plus making sure NOT to call force_reinit when initializing, otherwise it will overwrite your set_data.
This is quite an ugly workaround - it comes also with a warning when running :)

**Comment by zhreshold** [comment]
@sernaton You can have a constant param by get_constant: https://mxnet.incubator.apache.org/api/python/gluon/gluon.html?highlight=get_constant#mxnet.gluon.ParameterDict.get_constant

```python
filter_mask = ... # an nd.array
self.filter_mask = self.params.get_constant('filter_mask', filter_mask)
# no need to worry about initialization 
```
This is way more elegant than what you are doing right now.

**Comment by sernaton** [comment]
Thanks zhreshold, indeed I ended up using a constant for my use case. Much more elegant indeed, and it works too :).

Nevertheless still stands that it is currently not possible to initialize a "hybridized" parameter with a ndarray. I think this issue should stay open.

**Comment by roywei** [comment]
https://github.com/apache/incubator-mxnet/pull/12678#issuecomment-424966436 may have fixed it, will verify

**Comment by piyushghai** [comment]
@roywei Were you able to verify if the fix worked or not ?

**Comment by roywei** [comment]
@piyushghai  no, it should remain open. `self.params.get_constant` is recommend for now.
